src/hotspot/share/opto/loopnode.cpp

```diff
@@ -5232,6 +5253,68 @@ int PhaseIdealLoop::build_loop_tree_impl( Node *n, int pre_order ) {
   return pre_order;
 }
 
+#ifndef PRODUCT
+//--------------------------verify_regions_in_irreducible_loops----------------
+// Iterate down from Root through CFG, verify for every region:
+// if it is in an irreducible loop it must be marked as such
+void PhaseIdealLoop::verify_regions_in_irreducible_loops() {
+  ResourceMark rm;
+  //assert(_verify_only, "this verification is expensive");
+  if (!_has_irreducible_loops) {
+    // last build_loop_tree has not found any irreducible loops
+    // hence no region has to be marked is_in_irreduible_loop
+    return;
+  }
+
+  RootNode* root = Compile::current()->root();
+  Unique_Node_List worklist; // visit all nodes once
+  worklist.push(root);
+  bool failure = false;
+  for (uint i = 0; i < worklist.size(); i++) {
+    Node* n = worklist.at(i);
+    if (n->is_Region()) {
+      RegionNode* region = n->as_Region();
+      if (is_in_irreducible_loop(region) &&
+          region->loop_status() == RegionNode::LoopStatus::Reducible) {
+        failure = true;
+#ifdef ASSERT
+        tty->print("irreducible! ");
+        region->dump();
+#endif
+      }
+    }
+    for (DUIterator_Fast jmax, j = n->fast_outs(jmax); j < jmax; j++) {
+      Node* use = n->fast_out(j);
+      if (use->is_CFG()) {
+        worklist.push(use); // push if was not pushed before
+      }
+    }
+  }
+  assert(!failure, "region in irreducible loop was marked as reducible");
+}
+
+//---------------------------is_in_irreducible_loop-------------------------
+bool PhaseIdealLoop::is_in_irreducible_loop(RegionNode* region) {
+  if (!_has_irreducible_loops) {
+    return false; // no irreducible loop in graph
+  }
+  IdealLoopTree *l = get_loop(region); // l: innermost loop that contains region
+  do {
+    if (l->_irreducible) {
+      return true; // found it
+    }
+    if (l == _ltree_root) {
+      return false; // reached root, terimnate
+    }
+    l = l->_parent;
+  } while (l != nullptr);
+  // Infinite loops are not always properly attached to the loop-tree. Infinite
```

Comment:
@vnkozlov Should I change this too, just as in `ciTypeFlow::Block::is_in_irreducible_loop`?

Reply:
I now  did it for both, let me know if that is ok.